File: csharp/StrangeCharactersRefactoring/TestCharacters/ExampleCharacterTestCase.cs
using Characters;
using NUnit.Framework;

namespace TestCharacters;

/*
 * This test case shows what's possible with collection assertions
 */
public class ExampleCharacterTestCase
{
    [TestCase]
    public void OneCharacterTests()
    {
        // Arrange
        var nancy = new Character("Nancy", "Wheeler");
    
        // No Act step - we will just illustrate some assertions you can do on a Character
        
        // Assert
        Assert.AreEqual("Nancy", nancy.FirstName);
        Assert.AreEqual(nancy, nancy);
        // This assertion works because Character implements "Equals"
        Assert.AreEqual(new Character("Nancy", "Wheeler"), nancy);
    }
    
    [TestCase]
    public void CharacterListTests()
    {
        // Arrange
        var charactersList = CharacterFactory.WheelerFamily();
        
        // No Act step - we will just illustrate some assertions you can do on a List of Characters

        // Assert
        Assert.NotNull(charactersList);
        Assert.AreEqual(3, charactersList.Count);
        CollectionAssert.Contains(charactersList, new Character("Nancy", "Wheeler"));
        CollectionAssert.AreEquivalent(new List<Character>()
        {
            new Character("Nancy", "Wheeler"),
            new Character("Mike", "Wheeler"),
            new Character("Karen", "Wheeler"),
        }, charactersList);
        Assert.AreEqual(new List<Character>()
        {
            new Character("Karen", "Wheeler"),
            new Character("Mike", "Wheeler"),
            new Character("Nancy", "Wheeler"),
        }, charactersList);
    }
}

File: php/tests/ExampleCharacterTestCase.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use PHPUnit\Framework\Attributes\Test;
use StrangeCharacters\Character;
use StrangeCharacters\CharacterFactory;

class ExampleCharacterTestCase extends TestCase
{
    #[Test]
    public function oneCharacterTests(): void {
        // Arrange
        $nancy = Character::withFirstAndLastName("Nancy", "Wheeler");

        // No Act step - we will just illustrate some assertions you can do on a Character

        // Assert
        self::assertEquals("Nancy", $nancy->firstName);
        // This assertion works because Character implements "equals"
        self::assertEquals(Character::withFirstAndLastName("Nancy", "Wheeler"), $nancy);
    }

     #[Test]
     public function characterListTests() : void{
         // Arrange
         $charactersList = CharacterFactory::wheelerFamily();

         // No Act step - we will just illustrate some assertions you can do on a List of Characters

         // Assert
         self::assertNotNull($charactersList);
         self::assertCount(3, $charactersList);
         self::assertContains(Character::withFirstAndLastName("Nancy", "Wheeler"), $charactersList);
         self::assertEquals([
                 Character::withFirstAndLastName("Karen", "Wheeler"),
                 Character::withFirstAndLastName("Mike", "Wheeler"),
                 Character::withFirstAndLastName("Nancy", "Wheeler")
         ], array_values($charactersList));
         self::assertEquals([
                 "Karen Wheeler",
                 "Mike Wheeler",
                 "Nancy Wheeler"
         ], array_map(fn(Character $character) => $character->firstName. " " .$character->lastName,$charactersList));
     }
}
